lib/features/chat/services/chat_gpt_service.dart

```diff
@@ -1,51 +1,190 @@
-import 'package:chat_gpt_sdk/chat_gpt_sdk.dart';
-import 'package:cloud_firestore/cloud_firestore.dart';
-import 'package:flutter_chat_types/flutter_chat_types.dart' as types;
-import 'package:flutter_firebase_chat_core/flutter_firebase_chat_core.dart';
+import 'dart:convert';
 
-import '../../../resources/constants/chat_gpt_user.dart';
+import "package:chat_gpt_sdk/chat_gpt_sdk.dart";
+import "package:cloud_firestore/cloud_firestore.dart";
+import "package:flutter_chat_types/flutter_chat_types.dart" as types;
+import "package:flutter_firebase_chat_core/flutter_firebase_chat_core.dart";
+
+import "../../../resources/constants/chat_gpt_user.dart";
+import '../models/principle.dart';
+import '../models/principles.dart';
+import '../utils/extract_conversation.dart';
+import '../utils/get_prompt.dart';
+
+const _kLastMessageCount = 3;
 
 class ChatGptService {
-  static final openAI = OpenAI.instance.build(
+  static final _openAI = OpenAI.instance.build(
     token: '<< REPLACE WITH YOUR API KEY HERE>>',
     baseOption: HttpSetup(receiveTimeout: 6000),
     isLogger: true,
   );
 
-  static void sendReplyMessageFromChatGpt(
-    String prompt,
-    String roomId,
+  /// Sends a reply message from the GPT-3 chatbot to the specified room ID in Firebase.
```

Comment:
/// Sends a reply message from the GPT-3 chatbot to the specified room ID in Firebase.
  ///
  /// Given a recent message and a list of all messages in the room, the method will extract
  /// principles from the last two conversations, generate a prompt for the GPT-3 API using
  /// the extracted principles, and send the prompt to the API to generate a response. The
  /// response is then stored in Firestore and returned.
  ///
  /// Notes: This method will extract principles from the last two conversations and save them to
  /// Firestore for future use.